include/deal.II/base/lazy.h

```diff
@@ -0,0 +1,349 @@
+    // std::call_once.
+    //
+    reset_once_flag();
+
+    static_cast<std::optional<T> &>(*this) = other;
+
+    return *this;
+  }
+
+
+  /**
+   * Move assignment.
+   */
+  Lazy &
+  operator=(Lazy &&other) noexcept
+  {
+    //
+    // Move assignment might reset the std::optional<T> with an empty one
+    // containing no value. Thus, we have to reset the std::once_flag to
+    // "function not called".
+    //
+    // This is precisely what we want. We always check whether
+    // std::optional<T> contains a valid object first, and only if it
+    // doesn't we need the std::once_flag (in the *uninitialized state*) to
+    // initialize the std::optional<T> object with a call to
+    // std::call_once.
+    //
+    reset_once_flag();
+
+    static_cast<std::optional<T> &>(*this) = other;
+
+    return *this;
+  }
+
+
+  /**
+   * Swap the Lazy<T> object with another one.
+   */
+  void
+  swap(Lazy<T> &other) noexcept(
+    std::is_nothrow_move_constructible_v<T> &&std::is_nothrow_swappable_v<T>)
+  {
+    reset_once_flag();
+    other.reset_once_flag();
+
+    std::optional<T>::swap(other);
+  }
+
+
+#ifndef DOXYGEN
+  /**
+   * Disallow swapping with a plain std::optional<T> object.
+   */
+  void
+  swap(std::optional<T> &other) = delete;
+#endif
+
+
+  /**
+   * Reset the Lazy<T> object clearing the underlying std::optional<T>
+   * object.
+   */
+  void
+  reset() noexcept
+  {
+    reset_once_flag();
+    this->std::optional<T>::reset();
+  }
+
+
+#ifndef DOXYGEN
+  /**
+   * Disallow direct emplace().
+   */
+  template <class... Args>
+  T &
+  emplace(Args &&...) = delete;
+
+  /**
+   * Disallow direct emplace().
+   */
+  template <class U, class... Args>
+  T &
+  emplace(std::initializer_list<U>, Args &&...args) = delete;
+#endif
+
+
+  /**
+   * Initialize the wrapped object.
+   *
+   * If the contained object is already initialized this function simply
+   * returns and does nothing.
+   *
+   * If, instead, the object has not yet been initialized then the @p
+   * creator function object (oftentimes a lambda function) is called to initialize the contained object.
+   *
+   * This operation is thread safe: The ensure_initialized() method
+   * guarantees that the creator function object is only called once on one
+   * of the calling threads and that after completion the initialization
+   * result (which is stored in the std::optional) is visible on all
+   * threads.
+   */
+  template <typename Callable>
+  DEAL_II_ALWAYS_INLINE inline void
+  ensure_initialized(const Callable &creator)
+  {
+    if (!this->has_value())
+#ifdef DEAL_II_HAVE_CXX20
+      [[unlikely]]
+#endif
+      {
+        // We use std::call_once here to ensure that we call creator()
+        // exactly once on one thread and that the result (creation of
+        // object) is visible on all threads.
+        std::call_once(flag, [&]() {
+          static_cast<std::optional<T> &>(*this) = std::move(creator());
+        });
+      }
```

Comment:
I want to come back to this. I still haven't quite convinced myself that this is thread-safe. It relies on the fact that the operation
```
  static_cast<std::optional<T> &>(*this) = std::move(creator());
```
leads atomically to a change in observable `this->has_value()` behavior. But imagine that `std::optional::operator=(T&&)` is implemented as
```
  template <typename T>
  class optional {
    bool __has_value;
    char __value[sizeof(T)];
  public:
    optional& std::optional::operator=(T&& t) {
      __has_value = true;
      new(&value[0]) T(std::move(t));
      return *this;
    }
 };
```
This is not an unreasonable implementation. But it is not thread-safe. If one thread is right in the middle between the first and second statements of `operator=` here, then another thread might already see that `this->has_value()==true`, exit `ensure_initialized()`, and access the value via `std::optional::value()` before it is actually set.

Reply:
@bangerth That's a good point. Both implementation I checked actually implement the operations the other way round (make sure everything is in place, then flip the bool). BUT - we still don't have a guarantee that the memory store from the move and the change from the bool arrive in that order.

I will remove the std::optional entirely and replace it by a std::atomic<bool> to have acquire and release semantics.

Reply:
@bangerth In hindsight the old code that used `std::optional<T>::has_value()` would have been OK on amd64 (and with libstdc++ and libc++).

Reply:
I think that the `std::optional` was actually a good idea because it handles the memory issues dealing with a possibly uninitialized object. I just don't think that you can skip the `std::call_once`. You have to rely on `std::call_once` having a short exit path (which I expect all implementations to have), rather than implementing the short exit yourself on a separate variable.